node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,213 @@
+        G1::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode signature: {e:?}"))
+    }
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+/// Type safety wrapper around [Signature] indicating that it is an aggregated signature.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AggregateSignature(pub G1);
+
+impl AggregateSignature {
+    /// Generates an aggregate signature from a list of signatures.
+    pub fn aggregate<'a>(sigs: impl IntoIterator<Item = &'a Signature>) -> Self {
+        let mut agg = G1::zero();
+        for sig in sigs {
+            agg += sig.0
+        }
+
+        AggregateSignature(agg)
+    }
+
+    #[inline(never)]
+    fn verify_raw(&self, msgs_and_pks: &[(&[u8], &PublicKey)]) -> Result<(), Error> {
+        // Aggregate public keys if they are signing the same hash. Each public key aggregated
+        // is one fewer pairing to calculate.
+        let mut pairs: HashMap<&[u8], G2> = HashMap::new();
+        for (msg, pk) in msgs_and_pks {
+            *pairs.entry(msg).or_default() += pk.0;
+        }
+        // First pair: e(sig: G1, generator: G2)
+        let a = Bn254::pairing(self.0, G2::generator());
+
+        // Second pair: e(H(m1): G1, pk1: G2) * ... * e(H(m1000): G1, pk1000: G2)
+        let mut b = PairingOutput::zero();
+        for (msg, pk) in pairs {
+            b += Bn254::pairing(hash::hash_to_g1(msg), pk);
+        }
+
+        if a == b {
+            Ok(())
+        } else {
+            Err(Error::AggregateSignatureVerificationFailure)
+        }
+    }
+
+    /// Verifies an aggregated signature for multiple messages against the provided list of public keys.
+    /// This method expects one public key per message, otherwise it will fail. Note however that
+    /// If there are any duplicate messages, the public keys will be aggregated before verification.
+    pub fn verify<'a>(
+        &self,
+        msgs_and_pks: impl Iterator<Item = (&'a [u8], &'a PublicKey)>,
+    ) -> Result<(), Error> {
+        self.verify_raw(&msgs_and_pks.collect::<Vec<_>>()[..])
+    }
+}
+
+impl ByteFmt for AggregateSignature {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G1::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode aggregate signature: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl PartialOrd for AggregateSignature {
```

Comment:
Off-topic(?): Is `PartialOrd` / `Ord` required only for `BTreeMap`s in `PrepareQC`? Could `HashMap` be used there? It's not quite obvious why signatures should be ordered semantically.

Reply:
> Is `PartialOrd` / `Ord` required only for `BTreeMap`s in `PrepareQC`? 

Yes.

>  Could `HashMap` be used there? It's not quite obvious why signatures should be ordered semantically.

I didn't find use case for the current type of ordering. @pompon0 @brunoffranca any thoughts?

Reply:
We need reserialization of signable messages (like PrepareQC) to be an identity, so that signatures are valid at all times. This is kind of constraining, but the alternative would be to pass around the serialized message at all times so that reserialization is never needed (which is in fact more bullet-proof, but a bit costly and makes message handling more complex).

Reply:
+1